geo/aws.go

```diff
@@ -0,0 +1,135 @@
+package geo
+
+import (
+	"encoding/json"
+	"errors"
+	"log"
+	"net"
+	"os"
+	"time"
+)
+
+const (
+	AWSGeoUrl  = "https://ip-ranges.amazonaws.com/ip-ranges.json"
+	AWSExpired = 7 * 24 * time.Hour
+)
+
+type AWSProvider struct {
+	Info       AWSGeoInfo
+	InfoFile   string
+	LastUpdate time.Time
+}
+
+type AWSGeoInfo struct {
+	SyncToken  string      `json:"syncToken"`
+	CreateDate string      `json:"createDate"`
+	Prefixes   []AWSIPInfo `json:"prefixes"`
+}
+
+type AWSIPInfo struct {
+	IPPrefix           string     `json:"ip_prefix"`
+	Region             string     `json:"region"`
+	Service            string     `json:"service"`
+	NetworkBorderGroup string     `json:"network_border_group"`
+	Subnet             *net.IPNet `json:"-"`
+}
+
+var AWSRegionMapping = map[string]string{
+	"us-east-1":      "US",
+	"us-east-2":      "US",
+	"us-west-1":      "US",
+	"us-west-2":      "US",
+	"af-south-1":     "ZA",
+	"ap-east-1":      "HK",
+	"ap-south-1":     "IN",
+	"ap-northeast-1": "JP",
+	"ap-northeast-2": "KR",
+	"ap-northeast-3": "JP",
+	"ap-southeast-1": "SG",
+	"ap-southeast-2": "AU",
+	"ca-central-1":   "CA",
+	"eu-central-1":   "DE",
+	"eu-west-1":      "IE",
+	"eu-west-2":      "GB",
+	"eu-west-3":      "FR",
+	"eu-south-1":     "IT",
+	"eu-north-1":     "SE",
+	"me-south-1":     "BH",
+	"sa-east-1":      "BR",
+}
+
+func NewAWSProvider(AWSFile string) (*AWSProvider, error) {
+	if len(AWSFile) == 0 {
+		return nil, errors.New("invalid AWSFile path")
+	}
+	b, err := GetJsonFile(AWSGeoUrl, AWSFile)
+	if err != nil {
+		return nil, err
+	}
+	var info AWSGeoInfo
+	err = json.Unmarshal(b, &info)
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range info.Prefixes {
+		if len(p.IPPrefix) == 0 {
+			continue
+		}
+		_, subnet, err := net.ParseCIDR(p.IPPrefix)
+		if err != nil {
+			log.Print(err)
+			continue
+		}
+		p.Subnet = subnet
+	}
+
+	aws := new(AWSProvider)
+	aws.Info = info
+	aws.InfoFile = AWSFile
+	aws.LastUpdate = time.Now()
+
+	return aws, nil
+}
+
+func (p *AWSProvider) GetLocation(ip string) (*Location, error) {
+	if time.Since(p.LastUpdate) > AWSExpired {
+		err := os.Remove(p.InfoFile)
```

Comment:
Remove the old file before upgrade success is not a good way. We need to make sure if upgrade fails, old file can still be used.